r/R/dplyr-funcs-type.R

```diff
@@ -82,44 +82,40 @@ register_bindings_type_cast <- function() {
                                        tryFormats = "%Y-%m-%d",
                                        origin = "1970-01-01",
                                        tz = "UTC") {
+    binding_as_date(
+      x = x,
+      format = format,
+      tryFormats = tryFormats,
+      origin = origin,
+      tz = tz,
+      base = TRUE
+    )
+  })
 
-    # the origin argument will be better supported once we implement temporal
-    # arithmetic (https://issues.apache.org/jira/browse/ARROW-14947)
-    # TODO revisit once the above has been sorted
-    if (call_binding("is.numeric", x) & origin != "1970-01-01") {
-      abort("`as.Date()` with an `origin` different than '1970-01-01' is not supported in Arrow")
-    }
-
-    # this could be improved with tryFormats once strptime returns NA and we
-    # can use coalesce - https://issues.apache.org/jira/browse/ARROW-15659
-    # TODO revisit once https://issues.apache.org/jira/browse/ARROW-15659 is done
-    if (is.null(format) && length(tryFormats) > 1) {
-      abort("`as.Date()` with multiple `tryFormats` is not supported in Arrow")
-    }
-
-    if (call_binding("is.Date", x)) {
-      return(x)
-
-    # cast from POSIXct
-    } else if (call_binding("is.POSIXct", x)) {
-      # base::as.Date() first converts to the desired timezone and then extracts
-      # the date, which is why we need to go through timestamp() first
-      x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
-
-    # cast from character
-    } else if (call_binding("is.character", x)) {
-      format <- format %||% tryFormats[[1]]
-      # unit = 0L is the identifier for seconds in valid_time32_units
-      x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
+  register_binding("as_date", function(x,
+                                       format = NULL,
+                                       origin = "1970-01-01",
+                                       tz = "UTC") {
+    binding_as_date(
+      x = x,
+      format = format,
+      origin = origin,
+      tz = tz,
+      base = FALSE
+    )
+  })
 
-    # cast from numeric
-    } else if (call_binding("is.numeric", x) & !call_binding("is.integer", x)) {
-      # Arrow does not support direct casting from double to date32()
-      # https://issues.apache.org/jira/browse/ARROW-15798
-      # TODO revisit if arrow decides to support double -> date casting
-      abort("`as.Date()` with double/float is not supported in Arrow")
+  register_binding("as_datetime", function(x,
+                                           origin = "1970-01-01",
+                                           tz = "UTC") {
```

Comment:
I would think `as_date` accepts any `tz` and `as_datetime` has a default `tz` set to UTC?

Also, how come `as_datetime` doesn't accept `format` - I am wondering how come you did not use `binding_as_date` for `as_datetime` also?

Reply:
* I would think `as_date` accepts any `tz` and `as_datetime` has a default tz set to UTC?
**Correct**: the default value for `tz` in `as_datetime()` is `"UTC"` & `as_date()` accepts any `tz`. I changed the default value for `tz` in `as_date()` to `NULL`.

* How come `as_datetime()` doesn't accept `format`?
Oversight, I added support for `format`

* Why I didn't use `binding_as_date()` for `as_datetime`? 
`x` for `as_datetime()` might have a `time` component and `binding_as_date()` won't parse it.

Reply:
Done